rockcraft/models/project.py

```diff
@@ -110,41 +72,56 @@ def _validate_platform_set(cls, values: Mapping[str, Any]) -> Mapping[str, Any]:
 DEPRECATED_COLON_BASES = ["ubuntu:20.04", "ubuntu:22.04"]
 
 
-class NameStr(pydantic.ConstrainedStr):
-    """Constrained string type only accepting valid rock names."""
+ProjectName = Annotated[
+    str,
+    pydantic.BeforeValidator(
+        get_validator_by_regex(_PROJECT_NAME_COMPILED_REGEX, MESSAGE_INVALID_NAME)
+    ),
+    pydantic.Field(
+        min_length=1,
+        max_length=40,
+        strict=True,
+        pattern=PROJECT_NAME_REGEX,
+        description=_PROJECT_NAME_DESCRIPTION,
+        title="Project Name",
+    ),
+]
 
-    regex = re.compile(NAME_REGEX)
+BuildBaseT = typing.Annotated[
+    Literal["ubuntu@20.04", "ubuntu@22.04", "ubuntu@24.04", "devel"] | None,
+    pydantic.Field(validate_default=True),
+]
 
 
 class BuildPlanner(BaseBuildPlanner):
     """BuildPlanner for Rockcraft projects."""
 
     platforms: dict[str, Platform]  # type: ignore[assignment]
     base: Literal["bare", "ubuntu@20.04", "ubuntu@22.04", "ubuntu@24.04"]  # type: ignore[reportIncompatibleVariableOverride]
-    build_base: Literal["ubuntu@20.04", "ubuntu@22.04", "ubuntu@24.04", "devel"] | None  # type: ignore[reportIncompatibleVariableOverride]
+    build_base: BuildBaseT = None  # type: ignore[reportIncompatibleVariableOverride]
```

Comment:
This could also be:

```suggestion
    build_base: Literal["ubuntu@20.04", "ubuntu@22.04", "ubuntu@24.04", "devel"] = pydantic.Field(default=None, validate_default=True)  # type: ignore[reportIncompatibleVariableOverride]
```

I prefer having it separate if we can have a `BaseT` near it or de-duplicate the `ubuntu@` bases